### PR TITLE
feat!: Unify ABIs between nargo and yarn-project

### PR DESCRIPTION
**Pull request by spalladino**
Subrepo push of https://github.com/AztecProtocol/aztec-packages/pull/3989 to a branch in order to trigger Noir CI

Should be merged as part of a proper git subrepo push

### CODE REVIEW

File: compiler/wasm/src/compile.rs [hunk 1]
```diff
@@ -148,8 +140,8 @@ impl PathToFileSourceMap {
 }
 
 pub enum CompileResult {
-    Contract { contract: ContractArtifact, debug: DebugArtifact },
-    Program { program: ProgramArtifact, debug: DebugArtifact },
+    Contract { contract: ContractArtifact, warnings: Vec<String> },
+    Program { program: ProgramArtifact, warnings: Vec<String> },
```
Comment (TomAFrench) [inline compiler/wasm/src/compile.rs:144]: I realise now that we're actually dropping all of the warnings from the `noirc_frontend`. Shall we drop these (warnings from `noirc_evaluator`) as well for now and then we can add a `warnings` field which contains both of these?

Comment (TomAFrench) [reply compiler/wasm/src/compile.rs:144]: On third thoughts, I'm fairly sure that the types as emitted from the wasm are now completely ignored. There's types in the TS code which are what are meaningful now.

Comment (TomAFrench) [reply compiler/wasm/src/compile.rs]: Ideally thought we'd still surface these types from the wasm into the TS. Seems like the lazy loading is throwing a spanner in the works however.

**Comment by spalladino** [comment]
The error firing in the `codegen-verify` of the integration test seems to happen because of a wrong bb version being used. I could reproduce locally, then re-built bb and pointed `NARGO_BACKEND_PATH` to it, and code generation worked fine. Still, I did get an unreachable error running the recursion test:

```
test/browser/recursion.test.ts:

 🚧 Browser logs:
      Compiling at /src/main.nr
      Dependencies: 
      Adding source /src/main.nr
      Compiling at /src/main.nr
      Dependencies: 
      Adding source /src/main.nr

 ❌ It compiles noir program code, receiving circuit bytes and abi object. > Should generate valid inner proof for correct input, then verify proof within a proof
      RuntimeError: unreachable
        at wasm://wasm/017523ee:wasm-function[14038]:0x588099
        at wasm://wasm/017523ee:wasm-function[29]:0x1e831
        at wasm://wasm/017523ee:wasm-function[3889]:0x1aa7f3
        at wasm://wasm/017523ee:wasm-function[3887]:0x1aa4f9
        at wasm://wasm/017523ee:wasm-function[3885]:0x1aa470
        at wasm://wasm/017523ee:wasm-function[3884]:0x1aa402
        at wasm://wasm/017523ee:wasm-function[3844]:0x1a9830
        at wasm://wasm/017523ee:wasm-function[3666]:0x1a8162
        at wasm://wasm/017523ee:wasm-function[3652]:0x1a7c11
        at wasm://wasm/017523ee:wasm-function[3643]:0x1a795d
```

Not sure how to unblock this one.

File: compiler/wasm/src/compile.rs (hunk 1 above)
Comment (spalladino) [reply compiler/wasm/src/compile.rs]: See https://github.com/AztecProtocol/aztec-packages/pull/3989#issuecomment-1892720070

File: compiler/wasm/src/types/noir_artifact.ts [hunk 2]
```diff
@@ -57,12 +59,16 @@ export interface NoirFunctionEntry {
 export interface CompiledContract {
```
Comment (TomAFrench) [inline compiler/wasm/src/types/noir_artifact.ts:59]:
```suggestion
export interface ContractArtifact {
```

File: compiler/wasm/src/types/noir_artifact.ts [hunk 3]
```diff
@@ -57,12 +59,16 @@ export interface NoirFunctionEntry {
...
   /** The name of the contract. */
   name: string;
+  /** Version of noir used for the build. */
+  noir_version: string;
   /** Compilation backend. */
   backend: string;
```
Comment (TomAFrench) [inline compiler/wasm/src/types/noir_artifact.ts:65]:
```suggestion
```

File: compiler/wasm/src/types/noir_artifact.ts [hunk 4]
```diff
@@ -73,12 +79,14 @@ export interface CompiledCircuit {
   hash?: number;
   /** Compilation backend. */
   backend: string;
```
Comment (TomAFrench) [inline compiler/wasm/src/types/noir_artifact.ts:81]:
```suggestion
```

**Review by TomAFrench — COMMENTED** [review]
Looks fine to me. The JS types are still outdated and referring to internal compiler structs however.

**Comment by TomAFrench** [comment]
Closing as this got merged in aztec-packages.

**Pull request closed without merging**